### PR TITLE
feat(runner): Add staff and superadmin argument for create user

### PR DESCRIPTION
**Pull request by rgibert**
Add argument to be able to disable staff for superusers, for use in single tenant.

### CODE REVIEW

**Review by beezz — APPROVED** [review]
🚀

**Comment by rgibert** [comment]
Added a superadmin flag and cleaned up the handling the prompting/for superadmin.  Needed to be able to automate single tenant setup/initial user on-boarding.

File: src/sentry/runner/commands/createuser.py [hunk 1]
```diff
@@ -68,23 +80,31 @@ def createuser(email, password, superuser, no_password, no_input, force_update):
         if superuser is None:
             superuser = _get_superuser()
 
+        # for self hosted to give superusers admin permissions
+        if superuser and settings.SENTRY_SELF_HOSTED and superadmin is None:
+            superadmin = _get_superadmin()
+
+        if staff is None:
+            staff = _get_staff()
```
Comment (BYK) [inline src/sentry/runner/commands/createuser.py:88]: You should still check for `no_input` and not prompt for these if it is `True`?

Comment (rgibert) [reply src/sentry/runner/commands/createuser.py:88]: All these ifs are covered by the check herehttps://github.com/getsentry/sentry/pull/32802/files#diff-4c81db7e7d333c1766ce5cac76b775aae663f2f7ea0744fa44d75c92855b4331R73

**Comment by rgibert** [comment]
Also referencing https://github.com/getsentry/self-hosted/issues/1341 which this should fix I believe.

File: src/sentry/runner/commands/createuser.py (hunk 1 above)
Comment (BYK) [reply src/sentry/runner/commands/createuser.py:88]: ah, sorry missed that 😅

File: src/sentry/runner/commands/createuser.py [hunk 2]
```diff
@@ -37,27 +37,39 @@ def _get_superuser():
...
 @click.option("--password")
 @click.option("--superuser/--no-superuser", default=None, is_flag=True)
+@click.option("--superadmin/--no-superadmin", default=False, is_flag=True)
+@click.option("--staff/--no-staff", default=None, is_flag=True)
 @click.option("--no-password", default=False, is_flag=True)
 @click.option("--no-input", default=False, is_flag=True)
 @click.option("--force-update", default=False, is_flag=True)
 @configuration
-def createuser(email, password, superuser, no_password, no_input, force_update):
+def createuser(email, password, superuser, superadmin, staff, no_password, no_input, force_update):
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/createuser.py:68]: What's the permissions hierarchy between superuser, superadmin, staff? I suggest consistently ordering them from least to greatest permissions (or reverse, but consistently).

File: src/sentry/runner/commands/createuser.py [hunk 3]
```diff
@@ -68,23 +80,31 @@ def createuser(email, password, superuser, no_password, no_input, force_update):
         if superuser is None:
             superuser = _get_superuser()
 
+        # for self hosted to give superusers admin permissions
+        if superuser and settings.SENTRY_SELF_HOSTED and superadmin is None:
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/createuser.py:84]: How is it possible for `superadmin` to be `None` here in normal usage? The click default is `False`, and specifying the flag explicitly one way or another will likewise result in a boolean.

File: src/sentry/runner/commands/createuser.py [hunk 4]
```diff
@@ -68,23 +80,31 @@ def createuser(email, password, superuser, no_password, no_input, force_update):
...
+
+        if staff is None:
+            staff = _get_staff()
+
     if superuser is None:
         superuser = False
 
+    if staff is None:
+        staff = superuser
+
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/createuser.py:95]: I think we also will need a `if superadmin is None:` block here if we end up going with None-able `--superadmin`.

File: src/sentry/runner/commands/createuser.py [hunk 5]
```diff
@@ -37,27 +37,39 @@ def _get_superuser():
...
+    role = UserRole.objects.get(name="Super Admin")
+    UserRoleUser.objects.create(user=user, role=role)
 
 
 @click.command()
 @click.option("--email")
 @click.option("--password")
 @click.option("--superuser/--no-superuser", default=None, is_flag=True)
+@click.option("--superadmin/--no-superadmin", default=False, is_flag=True)
```
[the remark this answers is not in the document]
Comment (chadwhitacre) [reply src/sentry/runner/commands/createuser.py:62]: `_get_superuser` prompt says "This grants them all permissions available." But then what is the `staff` role about? 🤔

File: src/sentry/runner/commands/createuser.py [hunk 6]
```diff
@@ -37,27 +37,39 @@ def _get_superuser():
...
+@click.option("--staff/--no-staff", default=None, is_flag=True)
```
[the remark this answers is not in the document]
Comment (chadwhitacre) [reply src/sentry/runner/commands/createuser.py:63]: Based on further reading I think it makes the most sense to use `None` as the default for `--superadmin` to match the behavior of the other two role/permission switches (`--superuser` and `--staff`).

**Review by chadwhitacre — CHANGES_REQUESTED** [review]
Circling back from call w/ @rgibert ... todo:

- [x] move envvar seatbelt to top of function, limit to `SELF_HOSTED or SINGLE_TENANT`
- [x] remove `--superadmin` from config options, _always_ grant "Super Admin" role when `superuser is True`
- [x] preserve existing behavior when `--staff` is `None` (shadow `superuser`)
- [x] add a comment re: superuser, superadmin, and staff

File: src/sentry/runner/commands/createuser.py [hunk 7]
```diff
@@ -58,95 +51,97 @@ def _set_superadmin(user):
...
 @click.option("--no-input", default=False, is_flag=True)
-@click.option("--force-update", default=False, is_flag=True)
+@click.option("--force-update", default=False, is_flag=True, help="If true, will update existing users.")
 @configuration
-def createuser(email, password, superuser, superadmin, staff, no_password, no_input, force_update):
+def createuser(email, password, superuser, staff, no_password, no_input, force_update):
     "Create a new user."
 
-    from django.conf import settings
+    if settings.SENTRY_SELF_HOSTED or settings.SENTRY_SINGLE_ORGANIZATION:
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/createuser.py]: nit: I like a negative here for an early exit, clear short-circuit and avoids excessive indentation below.

```suggestion
    if not (settings.SENTRY_SELF_HOSTED or settings.SENTRY_SINGLE_ORGANIZATION):
        return
```

File: src/sentry/runner/commands/createuser.py [hunk 8]
```diff
@@ -37,97 +37,124 @@ def _get_superuser():
...
+                if len(teams) == 1:
+                    OrganizationMemberTeam.objects.create(team=teams[0], organizationmember=member)
+                click.echo(f"Added to organization: {org.slug}")
+
+        if password:
+            user.set_password(password)
+            user.save()
+
+        if superuser:
+            _set_superadmin(user)
```
Comment (chadwhitacre) [inline src/sentry/runner/commands/createuser.py:158]: I'd like to see a comment here or in the `_set_superadmin` function documenting that superadmin is a kludge that approximates `superuser` using the role system vs. a model attribute.

**Review by chadwhitacre — APPROVED** [review]
Approving with a couple nits.

File: src/sentry/runner/commands/createuser.py [hunk 9]
```diff
@@ -37,27 +37,46 @@ def _get_superuser():
...
+)
 @configuration
-def createuser(email, password, superuser, no_password, no_input, force_update):
+def createuser(email, password, superuser, staff, no_password, no_input, force_update):
     "Create a new user."
+
+    from django.conf import settings
+
+    if not (settings.SENTRY_SELF_HOSTED or settings.SENTRY_SINGLE_ORGANIZATION):
+        return
```
Comment (wedamija) [inline src/sentry/runner/commands/createuser.py:78]: @rgibert Just want to understand the context on this change. At the moment, this is causing `createuser` to fail on `getsentry`.

Comment (chadwhitacre) [reply src/sentry/runner/commands/createuser.py:78]: @wedamija @rgibert Did this get resolved somehow? Link?

Comment (chadwhitacre) [reply src/sentry/runner/commands/createuser.py:78]: https://github.com/getsentry/sentry/pull/37630 I think?

**Comment by rgibert** [comment]
We moved the settings check to the _set_superadmin() call which aligns with how it was working before this PR.

See https://github.com/getsentry/sentry/pull/37630

File: src/sentry/runner/commands/createuser.py (hunk 9 above)
Comment (chadwhitacre) [reply src/sentry/runner/commands/createuser.py:78]: [Yup](https://github.com/getsentry/sentry/pull/32802#issuecomment-1215143946). 👍

**Pull request merged**